### PR TITLE
refactor(parser): add ast types to parser

### PR DESCRIPTION
**Pull request by armano2**

### CODE REVIEW

File: packages/typescript-estree/src/parser.ts [hunk 1]
```diff
@@ -417,3 +413,5 @@ export function parseAndGenerateServices<
 
 export { AST_NODE_TYPES } from './ast-node-types';
 export { ParserOptions };
+export { ParserServices };
+export { es };
```
Comment (armano2) [inline packages/typescript-estree/src/parser.ts]: i'm unsure what name we should use when we are exporting all of `typedefs`

File: packages/parser/src/analyze-scope.ts [hunk 2]
```diff
@@ -61,27 +61,29 @@ class PatternVisitor extends OriginalPatternVisitor {
     }
   }
 
-  ArrayPattern(node: any) {
+  ArrayPattern(node: es.ArrayPattern): void {
     node.elements.forEach(this.visit, this);
-    if (node.decorators) {
-      this.rightHandNodes.push(...node.decorators);
+    // TODO: there is no decorators in definition of ArrayPattern
+    if ((node as any).decorators) {
```
Comment (armano2) [inline packages/parser/src/analyze-scope.ts]: there is no test cases for this, and i'm unsure if its possible

File: packages/parser/src/analyze-scope.ts [hunk 3]
```diff
@@ -61,27 +61,29 @@ class PatternVisitor extends OriginalPatternVisitor {
...
   }
 
-  ObjectPattern(node: any) {
+  ObjectPattern(node: es.ObjectPattern): void {
     node.properties.forEach(this.visit, this);
-    if (node.decorators) {
-      this.rightHandNodes.push(...node.decorators);
+    // TODO: there is no decorators in definition of ObjectPattern
+    if ((node as any).decorators) {
+      this.rightHandNodes.push(...(node as any).decorators);
```
Comment (armano2) [inline packages/parser/src/analyze-scope.ts]: there is no test cases for this, and i'm unsure if its possible

File: packages/parser/src/analyze-scope.ts [hunk 4]
```diff
@@ -61,27 +61,29 @@ class PatternVisitor extends OriginalPatternVisitor {
...
     }
   }
 
-  ObjectPattern(node: any) {
+  ObjectPattern(node: es.ObjectPattern): void {
     node.properties.forEach(this.visit, this);
-    if (node.decorators) {
-      this.rightHandNodes.push(...node.decorators);
+    // TODO: there is no decorators in definition of ObjectPattern
+    if ((node as any).decorators) {
```
Comment (armano2) [inline packages/parser/src/analyze-scope.ts]: there is no test cases for this, and i'm unsure if its possible

File: packages/parser/src/analyze-scope.ts [hunk 5]
```diff
@@ -214,7 +220,7 @@ class Referencer extends OriginalReferencer {
       this.visit(node.superTypeParameters);
     }
     if (node.implements) {
-      this.visit(node.implements);
+      node.implements.forEach(this.visit, this);
```
Comment (armano2) [inline packages/parser/src/analyze-scope.ts]: this is actual issue, `implements` in classes is a array of nodes but `visit` expect that you provide node iteslf

i have to design a way to test visiting, scope analysis snapshots does not contain information about implements, its part of type "scope", and its not yet implemented 

see #19

**Comment by armano2** [comment]
i'm going to add some additional test cases

File: packages/typescript-estree/src/parser.ts (hunk 1 above)
Comment (armano2) [reply packages/typescript-estree/src/parser.ts]: @bradzacher you are expecting that exported name will be `TSESTree` in #120

should i change it here?

Comment (armano2) [reply packages/typescript-estree/src/parser.ts]: i updated it already 👍

File: packages/typescript-estree/src/parser.ts [hunk 6]
```diff
@@ -14,8 +14,8 @@ import ts from 'typescript';
 import convert from './ast-converter';
 import { convertError } from './convert';
 import { firstDefined } from './node-utils';
-import * as es from './typedefs';
-import { Extra, ParserOptions } from './parser-options';
+import * as TSESTree from './typedefs';
```
Comment (JamesHenry) [inline packages/typescript-estree/src/parser.ts]: I was going to suggest this change (es -> TSESTree) in your last PR but didn't want to hold it up over something so trivial.

One thing though, can you go back and make this usage consistent? We use `es.` in other places than the ones which have been updated here.

Maybe do a barrel style export of a named symbol for all, and then import from the barrel instead?

```ts
// barrel
import * as TSEStree from '...'
export { TSEStree }
```

Then refactoring might be easier in future?

File: packages/parser/src/analyze-scope.ts [hunk 7]
```diff
@@ -11,7 +11,7 @@ import {
   PatternVisitorCallback,
   PatternVisitorOptions
 } from 'eslint-scope/lib/options';
-import { Node } from 'estree';
+import { TSESTree as es } from '@typescript-eslint/typescript-estree';
```
Comment (JamesHenry) [inline packages/parser/src/analyze-scope.ts]: why the extra rename?

Comment (armano2) [reply packages/parser/src/analyze-scope.ts]: ohh i missed this one

Comment (armano2) [reply packages/parser/src/analyze-scope.ts]: and i'm terrible at naming stuff

File: packages/typescript-estree/src/parser.ts (hunk 6 above)
Comment (armano2) [reply packages/typescript-estree/src/parser.ts]: i renames all es to TSEStree, i'm unsure what editor you are using but in Intelij its 
`right click` -> `refactor` -> `rename` -> and its done xd

**Comment by JamesHenry** [comment]
You need to do the rename in each file though, you are creating multiple
names symbols for the same thing in different places.

With my suggestion you could do the same refactor step on one single symbol
which would be imported by name everywhere

**Comment by JamesHenry** [comment]
^ Whoops added the above comment via email and it didn’t link it to your comment above

File: packages/typescript-estree/src/parser.ts (hunk 6 above)
Comment (JamesHenry) [reply packages/typescript-estree/src/parser.ts]: Left a comment below

**Pull request merged**